File: 3455.shortest-matching-substring.cpp
#
# @lc app=leetcode id=3455 lang=cpp
#
# [3455] Shortest Matching Substring
#

# @lc code=start
class Solution {
public:
    int shortestMatchingSubstring(string s, string p) {
        // Handle special edge case
        if (p == "**") return 0;
        int n = s.size();
        int m = p.size();
        // Find star positions
        int star1 = p.find('*');
        int star2 = p.find('*', star1 + 1);
        string prefix = p.substr(0, star1);
        string middle = p.substr(star1 + 1, star2 - star1 - 1);
        string suffix = p.substr(star2 + 1);
        int minLen = n + 1;
        for (int i = 0; i + prefix.size() <= n; ++i) {
            // Check prefix
            if (s.substr(i, prefix.size()) != prefix) continue;
            int start = i + prefix.size();
            // Find middle match
            for (int j = start; j + middle.size() <= n; ++j) {
                if (s.substr(j, middle.size()) != middle) continue;
                int afterMiddle = j + middle.size();
                // Check suffix
                if (afterMiddle + suffix.size() > n) continue;
                if (s.substr(afterMiddle, suffix.size()) != suffix) continue;
                // Valid match from i to afterMiddle + suffix.size() - 1
                int len = (afterMiddle + suffix.size()) - i;
                if (len < minLen) minLen = len;
            }
        }
        return (minLen <= n) ? minLen : -1;
    }
};
# @lc code=end

File: 3455.shortest-matching-substring.go
#
# @lc app=leetcode id=3455 lang=golang
#
# [3455] Shortest Matching Substring
#
# @lc code=start
import "sort"

func shortestMatchingSubstring(s string, p string) int {
    n := len(p)
    slen := len(s)
    star := []int{}
    for i := 0; i < n; i++ {
        if p[i] == '*' {
            star = append(star, i)
        }
    }
    star1, star2 := star[0], star[1]
    pre := p[:star1]
    mid := p[star1+1:star2]
    suf := p[star2+1:]
    // Special case: '**'
    if n == 2 && p[0] == '*' && p[1] == '*' {
        return 0
    }
    // Helper to find all indices where sub appears in s
    findAllIdx := func(sub string) []int {
        res := []int{}
        if len(sub) == 0 {
            for i := 0; i <= slen; i++ {
                res = append(res, i)
            }
            return res
        }
        for i := 0; i+len(sub) <= slen; i++ {
            if s[i:i+len(sub)] == sub {
                res = append(res, i)
            }
        }
        return res
    }
    preIdx := findAllIdx(pre)
    midIdx := findAllIdx(mid)
    sufIdx := findAllIdx(suf)
    // Ensure these slices are sorted for binary search
    sort.Ints(preIdx)
    sort.Ints(midIdx)
    sort.Ints(sufIdx)
    minLen := -1
    // For each prefix index, find valid orderings
    for _, i := range preIdx {
        preEnd := i + len(pre)
        // mid must start at or after preEnd
        mi := sort.SearchInts(midIdx, preEnd)
        for ; mi < len(midIdx); mi++ {
            j := midIdx[mi]
            midEnd := j + len(mid)
            if j < preEnd { continue }
            // suf must start at or after midEnd
            si := sort.SearchInts(sufIdx, midEnd)
            for ; si < len(sufIdx); si++ {
                k := sufIdx[si]
                if k < midEnd { continue }
                sufEnd := k + len(suf)
                curLen := sufEnd - i
                if minLen == -1 || curLen < minLen {
                    minLen = curLen
                }
                break // Only need minimal suf for this (i,j)
            }
        }
    }
    return minLen
}
# @lc code=end